lib/path-design.md

```diff
@@ -0,0 +1,177 @@
+# Path library design
+
+This document documents why the `lib.path` library is designed the way it is.
+
+The purpose of this library is to process paths. It does not read files from the filesystem.
+It exists to support the native Nix path value type with extra functionality.
+
+Since the path value type implicitly imports paths from the "eval-time system" into the store,
+this library explicitly doesn't support build-time or run-time paths, including paths to derivations.
+
+Overall, this library works with two basic forms of paths:
+- Absolute paths are represented with the Nix path value type. Nix automatically normalises these paths.
+- Relative paths are represented with the string value type. This library normalises these paths as safely as possible.
+
+Notably absolute paths in a string value type are not supported, the use of the string value type for relative paths is only because the path value type doesn't support relative paths.
+
+This library is designed to be as safe and intuitive as possible, throwing errors when operations are attempted that would produce surprising results, and giving the expected result otherwise.
+
+This library is designed to work well as a dependency for the `lib.filesystem` and `lib.sources` library components. Contrary to these library components, `lib.path` is designed to not read any paths from the filesystem.
+
+This library makes only these assumptions about paths and no others:
+- `dirOf path` returns the path to the parent directory of `path`, unless `path` is the filesystem root, in which case `path` is returned
+  - There can be multiple filesystem roots: `p == dirOf p` and `q == dirOf p` does not imply `p == q`
+    - While there's only a single filesystem root in stable Nix, the [lazy trees PR](https://github.com/NixOS/nix/pull/6530) introduces [additional filesystem roots](https://github.com/NixOS/nix/pull/6530#discussion_r1041442173)
+- `path + ("/" + string)` returns the path to the `string` subdirectory in `path`
+  - If `string` contains no `/` characters, then `dirOf (path + ("/" + string)) == path`
+  - If `string` contains no `/` characters, then `baseNameOf (path + ("/" + string)) == string`
+- `path1 == path2` returns true only if `path1` points to the same filesystem path as `path2`
+
+Notably we do not make the assumption that we can turn paths into strings using `toString path`.
+
+## Design decisions
+
+Each subsection here contains a decision along with arguments and counter-arguments for (+) and against (-) that decision.
+
+### Leading dots for relative paths
+[leading-dots]: #leading-dots-for-relative-paths
+
+Context: Relative paths can have a leading `./` to indicate it being a relative path, this is generally not necessary for tools though
+
+Decision: Returned relative paths should always have a leading `./`
```

Comment:
Context and Decision remind me of a format of parliamentary motions. You could succinctly summarize the most important positive argument(s) in between to make it complete.

```suggestion
Observing: Relative paths can have a leading `./` to indicate it being a relative path, this is generally not necessary for tools though

Considering: Paths should be as explicit, consistent and unambiguous as possible.

Decision: Returned relative paths should always have a leading `./`
```

Might save a couple of `<details>` clicks and bring a bit of order to the reasoning.

Reply:
Applied this format to all decisions